From 2a4ee0bd6e5503b404ce60539aed480569c3558e Mon Sep 17 00:00:00 2001
From: Evgeny Karpov <evgeny@kmaps.co>
Date: Thu, 27 Nov 2025 21:23:03 +0100
Subject: [PATCH 15/16] Support relocation for weak references (#4) (#1)

This patch adds handling weak references in similar way how it is done in _bfd_coff_generic_relocate_section.

Signed-off-by: Evgeny Karpov <evgeny@kmaps.co>
---
 bfd/coff-aarch64.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/bfd/coff-aarch64.c b/bfd/coff-aarch64.c
index a860912d81c..ef390b46606 100644
--- a/bfd/coff-aarch64.c
+++ b/bfd/coff-aarch64.c
@@ -530,11 +530,30 @@ coff_pe_aarch64_relocate_section (bfd *output_bfd,
 
       h = obj_coff_sym_hashes (input_bfd)[symndx];
 
-      if (h && h->root.type == bfd_link_hash_defined)
+      if (h && (h->root.type == bfd_link_hash_defined
+	|| h->root.type == bfd_link_hash_defweak))
 	{
 	  sec = h->root.u.def.section;
 	  sym_value = h->root.u.def.value;
 	}
+      else if (h && h->root.type == bfd_link_hash_undefweak
+	&& h->symbol_class == C_NT_WEAK && h->numaux == 1)
+	{
+	  struct coff_link_hash_entry *h2 =
+	  h->auxbfd->tdata.coff_obj_data->sym_hashes
+	  [h->aux->x_sym.x_tagndx.u32];
+
+	  if (!h2 || h2->root.type == bfd_link_hash_undefined)
+	  {
+	    sec = bfd_abs_section_ptr;
+	    sym_value = 0;
+	  }
+	  else
+	  {
+	    sec = h2->root.u.def.section;
+	    sym_value = h2->root.u.def.value;
+	  }
+	}
       else
 	{
 	  sec = sections[symndx];
-- 
2.51.0

